spring-rabbit/src/main/java/org/springframework/amqp/rabbit/core/RabbitTemplate.java

```diff
@@ -584,27 +586,126 @@ public void setBeanFactory(BeanFactory beanFactory) throws BeansException {
 	 * {@code Order} and finally unordered.
 	 * @param beforePublishPostProcessors the post processor.
 	 * @since 1.4.2
+	 * @see #addBeforePublishPostProcessor(MessagePostProcessor)
 	 */
 	public void setBeforePublishPostProcessors(MessagePostProcessor... beforePublishPostProcessors) {
 		Assert.notNull(beforePublishPostProcessors, "'beforePublishPostProcessors' cannot be null");
 		Assert.noNullElements(beforePublishPostProcessors, "'beforePublishPostProcessors' cannot have null elements");
 		this.beforePublishPostProcessors = MessagePostProcessorUtils.sort(Arrays.asList(beforePublishPostProcessors));
 	}
 
+	/**
+	 * Add {@link MessagePostProcessor} that will be invoked immediately before invoking
+	 * {@code Channel#basicPublish()}, after all other processing, except creating the
+	 * {@link BasicProperties} from {@link MessageProperties}. May be used for operations
+	 * such as compression. Processors are invoked in order, depending on {@code PriorityOrder},
+	 * {@code Order} and finally unordered.
+	 * <p>
+	 * In contrast to {@link #setBeforePublishPostProcessors(MessagePostProcessor...)}, this
+	 * method does not override the previously added beforePublishPostProcessors.
+	 * @param beforePublishPostProcessor the post processor.
+	 * @since 2.1.4
+	 */
+	public void addBeforePublishPostProcessor(MessagePostProcessor beforePublishPostProcessor) {
+		Assert.notNull(beforePublishPostProcessor, "'beforePublishPostProcessor' cannot be null");
+		if (this.beforePublishPostProcessors == null) {
+			this.beforePublishPostProcessors = new ArrayList<>();
+		}
+		this.beforePublishPostProcessors.add(beforePublishPostProcessor);
+		this.beforePublishPostProcessors = MessagePostProcessorUtils.sort(this.beforePublishPostProcessors);
+	}
+
+	/**
+	 * Remove the provided {@link MessagePostProcessor} from the {@link #beforePublishPostProcessors} list.
+	 * @param beforePublishPostProcessor the MessagePostProcessor to remove.
+	 * @since 2.1.4
+	 * @see #addBeforePublishPostProcessor(MessagePostProcessor)
+	 */
+	public void removeBeforePublishPostProcessor(MessagePostProcessor beforePublishPostProcessor) {
+		Assert.notNull(beforePublishPostProcessor, "'beforePublishPostProcessor' cannot be null");
+		if (this.beforePublishPostProcessors != null) {
+			this.beforePublishPostProcessors.remove(beforePublishPostProcessor);
+		}
+	}
+
+	/**
+	 * Remove a {@link MessagePostProcessor} from the {@link #beforePublishPostProcessors} list for specific index.
+	 * @param index the index for the MessagePostProcessor to remove.
+	 * @return the removed MessagePostProcessor
+	 * @since 2.1.4
+	 * @see #addBeforePublishPostProcessor(MessagePostProcessor)
+	 */
+	@Nullable
+	public MessagePostProcessor removeBeforePublishPostProcessor(int index) {
+		if (this.beforePublishPostProcessors != null) {
+			return new ArrayList<>(this.beforePublishPostProcessors).remove(index);
```

Comment:
Doesn't look good. you remove from a newly created `ArrayList` meanwhile the original one is still going to an item on that position.